core/src/polling_monitor/mod.rs

```diff
@@ -0,0 +1,232 @@
+use std::collections::VecDeque;
+use std::fmt::Display;
+use std::sync::Arc;
+
+use futures::stream;
+use futures::stream::StreamExt;
+use graph::cheap_clone::CheapClone;
+use graph::parking_lot::Mutex;
+use graph::prelude::tokio;
+use graph::slog::{debug, Logger};
+use tokio::sync::{mpsc, watch};
+use tower::{Service, ServiceExt};
+
+pub mod ipfs_service;
+
+/// Spawn a monitor that actively polls a service. Whenever the service has capacity, the monitor
+/// pulls object ids from the queue and polls the service. If the object is not present or in case
+/// of error, the object id is pushed to the back of the queue to be polled again.
+///
+/// The service returns the request ID along with errors or responses. The response is an
+/// `Option`, to represent the object not being found.
+pub fn spawn_monitor<ID, S, E, Response: Send + 'static>(
+    service: S,
+    response_sender: mpsc::Sender<(ID, Response)>,
+    logger: Logger,
+) -> PollingMonitor<ID>
+where
+    ID: Display + Send + 'static,
+    S: Service<ID, Response = (ID, Option<Response>), Error = (ID, E)> + Send + 'static,
+    E: Display + Send + 'static,
+    S::Future: Send,
+{
+    let queue = Arc::new(Mutex::new(VecDeque::new()));
+    let (wake_up_queue, queue_woken) = watch::channel(());
+
+    let queue_to_stream = {
+        let queue = queue.cheap_clone();
+        stream::unfold((), move |()| {
+            let queue = queue.cheap_clone();
+            let mut queue_woken = queue_woken.clone();
+            async move {
+                loop {
+                    let id = queue.lock().pop_front();
+                    match id {
+                        Some(id) => break Some((id, ())),
+                        None => match queue_woken.changed().await {
+                            // Queue woken, check it.
+                            Ok(()) => {}
+
+                            // The `PollingMonitor` has been dropped, cancel this task.
+                            Err(_) => break None,
+                        },
+                    };
+                }
+            }
+        })
+    };
+
+    {
+        let queue = queue.cheap_clone();
+        graph::spawn(async move {
+            let mut responses = service.call_all(queue_to_stream).unordered().boxed();
+            while let Some(response) = responses.next().await {
+                match response {
```

Comment:
For each `PollingMonitor<S>`, I think it would make sense to have metrics for:
- queue depth
- timeout rate
- error rate